File: src/main/java/com/pedigree/storage/ProjectRepository.java
package com.pedigree.storage;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.pedigree.model.*;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.AtomicMoveNotSupportedException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.time.Instant;
import java.util.List;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import java.util.zip.ZipOutputStream;

public class ProjectRepository {
    public static class ProjectData {
        public List<Individual> individuals = new com.pedigree.util.DirtyObservableList<>();
        public List<Family> families = new com.pedigree.util.DirtyObservableList<>();
        public List<Relationship> relationships = new com.pedigree.util.DirtyObservableList<>();
        public List<Tag> tags = new com.pedigree.util.DirtyObservableList<>();
    }

    private final ObjectMapper objectMapper;

    public ProjectRepository() {
        this.objectMapper = new ObjectMapper();
        this.objectMapper.registerModule(new JavaTimeModule());
        // Prevent Jackson from closing underlying streams (Zip streams) to avoid "stream closed"
        this.objectMapper.configure(JsonGenerator.Feature.AUTO_CLOSE_TARGET, false);
        this.objectMapper.configure(JsonParser.Feature.AUTO_CLOSE_SOURCE, false);
    }

    public void write(Path path, ProjectData data, ProjectLayout layout, ProjectMetadata meta) throws IOException {
        // Ensure parent directory exists
        Path dir = path.toAbsolutePath().getParent();
        if (dir != null && !Files.exists(dir)) {
            Files.createDirectories(dir);
        }

        // Update metadata modified timestamp
        if (meta != null) {
            meta.setModifiedAt(Instant.now());
        }

        // Write to a temporary file first
        Path tmp = Files.createTempFile(dir != null ? dir : path.toAbsolutePath().getParent(), path.getFileName().toString() + ".", ".tmp");
        boolean success = false;
        try (OutputStream os = Files.newOutputStream(tmp);
             ZipOutputStream zos = new ZipOutputStream(os)) {
            writeEntry(zos, ProjectFormat.DATA_JSON, data);
            writeEntry(zos, ProjectFormat.LAYOUT_JSON, layout);
            writeEntry(zos, ProjectFormat.META_JSON, meta);
            success = true;
        } catch (IOException e) {
            Files.deleteIfExists(tmp);
            throw e;
        }

        // Only rotate backups and replace the target if we successfully wrote the temp file
        if (success) {
            // Rotate backups before replacing the project file
            BackupService.rotateBackups(path);
            try {
                Files.move(tmp, path, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
            } catch (AtomicMoveNotSupportedException ex) {
                // Fallback if filesystem doesn't support atomic moves
                Files.move(tmp, path, StandardCopyOption.REPLACE_EXISTING);
            }
            // Mark project as clean after successful save
            com.pedigree.util.DirtyFlag.clear();
        }
    }

    private void writeEntry(ZipOutputStream zos, String name, Object payload) throws IOException {
        ZipEntry entry = new ZipEntry(name);
        zos.putNextEntry(entry);
        objectMapper.writeValue(zos, payload);
        zos.closeEntry();
    }

    public LoadedProject read(Path path) throws IOException {
        ProjectData data = null;
        ProjectLayout layout = null;
        ProjectMetadata meta = null;

        try (InputStream is = Files.newInputStream(path);
             ZipInputStream zis = new ZipInputStream(is)) {
            ZipEntry entry;
            while ((entry = zis.getNextEntry()) != null) {
                switch (entry.getName()) {
                    case ProjectFormat.DATA_JSON -> data = objectMapper.readValue(zis, ProjectData.class);
                    case ProjectFormat.LAYOUT_JSON -> layout = objectMapper.readValue(zis, ProjectLayout.class);
                    case ProjectFormat.META_JSON -> meta = objectMapper.readValue(zis, ProjectMetadata.class);
                    default -> { /* ignore */ }
                }
                zis.closeEntry();
            }
        }
        com.pedigree.util.DirtyFlag.clear();
        // Ensure lists are dirty-observable after load
        if (data != null) {
            data.individuals = new com.pedigree.util.DirtyObservableList<>(data.individuals);
            data.families = new com.pedigree.util.DirtyObservableList<>(data.families);
            data.relationships = new com.pedigree.util.DirtyObservableList<>(data.relationships);
            data.tags = new com.pedigree.util.DirtyObservableList<>(data.tags);
        }
        com.pedigree.util.DirtyFlag.clear();
        return new LoadedProject(data, layout, meta);
    }

    public record LoadedProject(ProjectData data, ProjectLayout layout, ProjectMetadata meta) {}
}

File: src/main/kotlin/com/family/tree/storage/ProjectRepository.kt
package com.family.tree.storage

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import com.family.tree.model.ProjectLayout
import com.family.tree.model.ProjectMetadata
import java.io.IOException
import java.nio.file.AtomicMoveNotSupportedException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.time.Instant
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream

class ProjectRepository {
    class ProjectData {
        @JvmField var individuals: MutableList<com.family.tree.model.Individual> = com.family.tree.util.DirtyObservableList()
        @JvmField var families: MutableList<com.family.tree.model.Family> = com.family.tree.util.DirtyObservableList()
        @JvmField var relationships: MutableList<com.family.tree.model.Relationship> = com.family.tree.util.DirtyObservableList()
        @JvmField var tags: MutableList<com.family.tree.model.Tag> = com.family.tree.util.DirtyObservableList()
        @JvmField var sources: MutableList<com.family.tree.model.Source> = com.family.tree.util.DirtyObservableList()
        @JvmField var repositories: MutableList<com.family.tree.model.Repository> = com.family.tree.util.DirtyObservableList()
        @JvmField var submitters: MutableList<com.family.tree.model.Submitter> = com.family.tree.util.DirtyObservableList()
    }

    private val objectMapper: ObjectMapper = ObjectMapper()
        .registerModule(JavaTimeModule())
        .registerKotlinModule()
        .configure(JsonGenerator.Feature.AUTO_CLOSE_TARGET, false)
        .configure(JsonParser.Feature.AUTO_CLOSE_SOURCE, false)

    @Throws(IOException::class)
    fun write(path: Path, data: ProjectData?, layout: ProjectLayout?, meta: ProjectMetadata?) {
        // Ensure parent directory exists
        val dir = path.toAbsolutePath().parent
        if (dir != null && !Files.exists(dir)) {
            Files.createDirectories(dir)
        }

        // Update metadata modified timestamp
        if (meta != null) {
            meta.modifiedAt = Instant.now()
        }

        // Write to a temporary file first
        val tmp = Files.createTempFile(dir ?: path.toAbsolutePath().parent, path.fileName.toString() + ".", ".tmp")
        var success = false
        try {
            Files.newOutputStream(tmp).use { os ->
                ZipOutputStream(os).use { zos ->
                    writeEntry(zos, ProjectFormat.DATA_JSON, data)
                    writeEntry(zos, ProjectFormat.LAYOUT_JSON, layout)
                    writeEntry(zos, ProjectFormat.META_JSON, meta)
                    success = true
                }
            }
        } catch (e: IOException) {
            Files.deleteIfExists(tmp)
            throw e
        }

        if (success) {
            // Rotate backups before replacing the project file
            BackupService.rotateBackups(path)
            try {
                Files.move(tmp, path, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE)
            } catch (ex: AtomicMoveNotSupportedException) {
                // Fallback if filesystem doesn't support atomic moves
                Files.move(tmp, path, StandardCopyOption.REPLACE_EXISTING)
            }
            // Mark project as clean after successful save
            com.family.tree.util.DirtyFlag.clear()
        }
    }

    @Throws(IOException::class)
    private fun writeEntry(zos: ZipOutputStream, name: String, payload: Any?) {
        val entry = ZipEntry(name)
        zos.putNextEntry(entry)
        objectMapper.writeValue(zos, payload)
        zos.closeEntry()
    }

    @Throws(IOException::class)
    fun read(path: Path): LoadedProject {
        var data: ProjectData? = null
        var layout: ProjectLayout? = null
        var meta: ProjectMetadata? = null

        Files.newInputStream(path).use { `is` ->
            ZipInputStream(`is`).use { zis ->
                var entry: ZipEntry?
                while (true) {
                    entry = zis.nextEntry ?: break
                    when (entry!!.name) {
                        ProjectFormat.DATA_JSON -> data = objectMapper.readValue(zis, ProjectData::class.java)
                        ProjectFormat.LAYOUT_JSON -> layout = objectMapper.readValue(zis, ProjectLayout::class.java)
                        ProjectFormat.META_JSON -> meta = objectMapper.readValue(zis, ProjectMetadata::class.java)
                        else -> {}
                    }
                    zis.closeEntry()
                }
            }
        }
        com.family.tree.util.DirtyFlag.clear()
        // Ensure lists are dirty-observable after load
        if (data != null) {
            data!!.individuals = com.family.tree.util.DirtyObservableList(data!!.individuals)
            data!!.families = com.family.tree.util.DirtyObservableList(data!!.families)
            data!!.relationships = com.family.tree.util.DirtyObservableList(data!!.relationships)
            data!!.tags = com.family.tree.util.DirtyObservableList(data!!.tags)
            data!!.sources = com.family.tree.util.DirtyObservableList(data!!.sources)
            data!!.repositories = com.family.tree.util.DirtyObservableList(data!!.repositories)
            data!!.submitters = com.family.tree.util.DirtyObservableList(data!!.submitters)
        }
        com.family.tree.util.DirtyFlag.clear()
        return LoadedProject(data, layout, meta)
    }

    data class LoadedProject(val data: ProjectData?, val layout: ProjectLayout?, val meta: ProjectMetadata?)
}
